File: 0198-house-robber/0198-house-robber.cpp
class Solution {
public:
    int helper(vector<int>& nums, int n, vector<int>& dp) {
        if (n == 0)
            return nums[0];
        if (n < 0)
            return 0;
        if (dp[n] != -1)
            return dp[n];
        int left = helper(nums, n - 1, dp) + 0;
        int right = helper(nums, n - 2, dp) + nums[n];
        return dp[n] = max(left, right);
    }
    int rob(vector<int>& nums) {
        int n = nums.size();
        vector<int> dp(n, -1);

        return helper(nums, n - 1, dp);
    }
};

File: 0198-house-robber/0198-house-robber.java
class Solution {
   public int helper(int[] nums, int index, int[] dp) {
        if (index >= nums.length)
            return 0;
        if (dp[index] != -1)
            return dp[index];
        int pick = nums[index] + helper(nums, index + 2, dp);
        int notPick = helper(nums, index + 1, dp);
        return dp[index] = Math.max(pick, notPick);
    }

    public int rob(int[] nums) {
        int n = nums.length;
        int[] dp = new int[n];
        Arrays.fill(dp, -1);
        return helper(nums, 0, dp);
    }
}